core/src/main/java/feign/Promise.java

```diff
@@ -0,0 +1,25 @@
+package feign;
+
+public class Promise<T> {
```

Comment:
promise is a loaded term, although I get what you are doing. how about using an AtomicReference instead? that way, we don't need to test as much.